Modules/_heapqmodule.c

```diff
@@ -538,6 +538,460 @@ _heapq__heapify_max(PyObject *module, PyObject *heap)
+                    return NULL;
+            }
+        }
+        mo->started = 1;
+    }
+    if (mo->tree == NULL)
+        return NULL;
+    if (mo->keyfunc == NULL) {
+        if (_tree_sift(mo, 0) < 0)
+            return NULL;
+        result = PyList_GET_ITEM(mo->tree, 0);
+        PyList_SET_ITEM(mo->tree, 0, NULL);
+    }
+    else {
+        if (_tree_sift_key(mo, 0) < 0)
+            return NULL;
+        Py_XDECREF(PyList_GET_ITEM(mo->tree, 0));
+        PyList_SET_ITEM(mo->tree, 0, NULL);
+        result = PyList_GET_ITEM(mo->tree, 1);
+        PyList_SET_ITEM(mo->tree, 1, NULL);
+    }
+    return result;
+}
+
+
+PyDoc_STRVAR(sizeof_doc, "Returns size in memory, in bytes.");
+
+static PyMethodDef merge_methods[] = {
+    {"__sizeof__", (PyCFunction)merge_sizeof, METH_NOARGS, sizeof_doc},
+    {NULL, NULL}
+};
+
+PyDoc_STRVAR(merge_doc,
+"merge(*iterables, key=None, reverse=False) --> merge object\n\
+\n\
+Merge multiple sorted inputs into a single sorted output.\n\
+\n\
+Similar to sorted(itertools.chain(*iterables)) but returns an iterator,\n\
+does not pull the data into memory all at once, and assumes that each of\n\
+the input streams is already sorted (smallest to largest).\n\
+\n\
+>>> list(merge([1,3,5,7], [0,2,4,8], [5,10,15,20], [], [25]))\n\
+[0, 1, 2, 3, 4, 5, 5, 7, 8, 10, 15, 20, 25]\n\
+\n\
+If *key* is not None, applies a key function to each element to determine\n\
+its sort order.\n\
+\n\
+>>> list(merge(['dog', 'horse'], ['cat', 'fish', 'kangaroo'], key=len))\n\
+['dog', 'cat', 'fish', 'horse', 'kangaroo']");
+
+static PyTypeObject merge_type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+   "heapq.merge",                       /* tp_name */
+    sizeof(mergeobject),                /* tp_basicsize */
+    0,                                  /* tp_itemsize */
+    /* methods */
+    (destructor)merge_dealloc,          /* tp_dealloc */
+    0,                                  /* tp_vectorcall_offset */
+    0,                                  /* tp_getattr */
+    0,                                  /* tp_setattr */
+    0,                                  /* tp_as_async */
+    0,                                  /* tp_repr */
+    0,                                  /* tp_as_number */
+    0,                                  /* tp_as_sequence */
+    0,                                  /* tp_as_mapping */
+    0,                                  /* tp_hash */
+    0,                                  /* tp_call */
+    0,                                  /* tp_str */
+    PyObject_GenericGetAttr,            /* tp_getattro */
+    0,                                  /* tp_setattro */
+    0,                                  /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC |
+        Py_TPFLAGS_BASETYPE,            /* tp_flags */
+    merge_doc,                          /* tp_doc */
+    (traverseproc)merge_traverse,       /* tp_traverse */
+    0,                                  /* tp_clear */
```

Comment:
Please, implement tp_clear as the tp_clear slot (maybe we could prove that this object cannot participate in reference cycles but is risky and the implementation is straightforward and assures correctness).